### ISSUE TITLE
16 bit grayscale images support

### ISSUE THREAD

**Issue by drewhk** [kardeiz/jp2k#1]
Right now I see that decoding results in a struct with a `Vec<u8>`, `height`, `width` and `bands`. However it looks like there is no field for bit depth. Are only 8-bit images supported?

**Comment by kardeiz** [comment]
Hi @drewhk, I'm not an `openjpeg` or `jpeg2000` expert, and I really only tested this on a small collection of `jp2` images.

However, I just tried opening a 16-bit `jp2` file, and it seems to work fine. The `buffer` provided in `jp2k::ImageBuffer` is a raw data array with the following layout:

> The memory area [is] a simple array, for example RGBRGBRGB, left-to-right, top-to-bottom.
> &mdash; https://libvips.github.io/libvips/API/current/VipsImage.html#vips-image-new-from-memory

The exact pixel-related data (e.g `RGB`) depends on how many `components` the `jp2` image has.

Please let me know if you think this should do something else...

**Comment by kardeiz** [comment]
Hi @drewhk, I'm not an `openjpeg` or `jpeg2000` expert, and I really only tested this on a small collection of `jp2` images.

However, I just tried opening a 16-bit `jp2` file, and it seems to work fine. The `buffer` provided in `jp2k::ImageBuffer` is a raw data array with the following layout:

> The memory area [is] a simple array, for example RGBRGBRGB, left-to-right, top-to-bottom.
> &mdash; https://libvips.github.io/libvips/API/current/VipsImage.html#vips-image-new-from-memory

The exact pixel-related data (e.g `RGB`) depends on how many `components` the `jp2` image has.

Please let me know if you think this should do something else...

**Comment by drewhk** [comment]
Yes, I tried and it works, but since you cast to `u8` at some point the output is wrong (16 bit values overflow the byte). Right now I switched to `openjpeg-sys` directly as I need other tweaks as well...
